src/Apps/ArtAdvisor/06-Discovery-Neartext/weaviate-db.ts

```diff
@@ -0,0 +1,264 @@
+    if (user.likedArtworkIds.includes(artworkInternalID)) return false
+    if (user.dislikedArtworkIds.includes(artworkInternalID)) return false
+
+    // else go ahead and create
+
+    const artworkId = generateUuid5(artworkInternalID)
+
+    const referenceProperty = {
+      like: "likedArtworks",
+      dislike: "dislikedArtworks",
+    }[reaction]
+
+    await this.client.data
+      .referenceCreator()
+      .withClassName(this.userClass)
+      .withId(userId)
+      .withReferenceProperty(referenceProperty)
+      .withReference(
+        this.client.data
+          .referencePayloadBuilder()
+          .withClassName(this.artworkClass)
+          .withId(artworkId)
+          .payload()
+      )
+      .do()
+
+    return true
+  }
+
+  /**
+   * Fetch artworks by semantic search on an array of concepts,
+   * optionally with liked and disliked artworks to move to/away from.
+   */
+  async getArtworksNearConcepts({
+    concepts,
+    likedArtworkIds = [],
+    dislikedArtworkIds = [],
+    limit = 10,
+  }: {
+    /** List of concepts for semantic search */
+    concepts: string[]
+    /** List of liked artwork Mongo ids */
+    likedArtworkIds?: MongoID[]
+    /** List of liked artwork Mongo ids */
+    dislikedArtworkIds?: MongoID[]
+    /** Max number of artworks to return */
+    limit?: number
+  }) {
+    console.log("[WeaviateDB] getArtworksNearConcepts", {
+      concepts,
+      likedArtworkIds,
+      dislikedArtworkIds,
+      limit,
+    })
+
+    const conceptArray = ensureValidConcepts(concepts)
+
+    const response = await this.client.graphql
+      .get()
+      .withClassName(this.artworkClass)
+      .withNearText({
+        concepts: conceptArray,
+        moveTo: getMoveObjects(likedArtworkIds),
+        moveAwayFrom: getMoveObjects(dislikedArtworkIds),
+      })
+      .withLimit(limit)
+      .withFields(
+        "internalID slug title date rarity medium materials price dimensions imageUrl _additional { id distance }"
+      )
+      .do()
+
+    const result = response.data.Get.DiscoveryArtworks.map(artwork => {
+      const properties = _.pick(artwork, [
+        "internalID",
+        "slug",
+        "title",
+        "date",
+        "rarity",
+        "medium",
+        "materials",
+        "price",
+        "dimensions",
+        "imageUrl",
+      ])
+
+      return {
+        id: artwork._additional.id,
+        ...properties,
+        distance: artwork._additional.distance,
+      }
+    })
+
+    return result as DiscoveryArtwork[]
+  }
+}
+
+/**
+ * Ensure that the concepts are valid for nearText(concepts: […])
+ * queries, to avoid Weaviate errors.
+ *
+ * @returns a non-empty array of non-empty strings
+ */
+function ensureValidConcepts(concepts: string[]) {
+  let conceptsArray = _.castArray(concepts)
+  conceptsArray = _.reject(conceptsArray, _.isEmpty)
+
+  if (_.isEmpty(conceptsArray)) {
+    conceptsArray.push("art") // TODO: better default?
+  }
```

Comment:
Do we need to handle edge cases where we have an array of `[undefined]` or `[null]`? Specifically, could we end up with `[null, "art"]`, and would that blow up the `weaviate` search? May be overly cautioius, but came to mind as I reminded myself how `castArray` [works](https://lodash.com/docs/4.17.15#castArray).

Reply:
L246 should prevent that by filtering out null, undefined or the empty string.